clippy_lints/src/copy_refcell.rs

```diff
@@ -0,0 +1,112 @@
+use clippy_config::Conf;
+use rustc_hir::{FieldDef, LetStmt, LocalSource};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_middle::ty::layout::TyAndLayout;
+use rustc_session::impl_lint_pass;
+use rustc_span::symbol::sym;
+use rustc_span::Span;
+
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::ty::implements_trait;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Detects crate-local usage of `RefCell<T>` where `T` implements `Copy`
+    ///
+    /// ### Why is this bad?
+    ///
+    /// `RefCell` has to perform extra book-keeping in order to support references, whereas `Cell` does not.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let interior_mutable = std::cell::RefCell::new(0_u8);
+    ///
+    /// *interior_mutable.borrow_mut() = 1;
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let interior_mutable = std::cell::Cell::new(0_u8);
+    ///
+    /// interior_mutable.set(1);
+    /// ```
+    #[clippy::version = "1.83.0"]
+    pub COPY_REFCELL,
+    perf,
+    "usage of `RefCell<T>` where `T` implements `Copy`"
+}
+
+pub struct CopyRefcell {
+    maximum_size: u64,
+}
+
+impl CopyRefcell {
+    pub fn new(conf: &'static Conf) -> Self {
+        Self {
+            maximum_size: conf.large_cell_limit,
+        }
+    }
+
+    fn check_copy_refcell<'tcx>(&self, cx: &LateContext<'tcx>, ty: ty::Ty<'tcx>, span: Span) {
+        let Some(copy_def_id) = cx.tcx.get_diagnostic_item(sym::Copy) else {
+            return;
+        };
+
+        let ty::Adt(adt, generics) = ty.kind() else {
+            return;
+        };
+
+        if !cx.tcx.is_diagnostic_item(sym::RefCell, adt.did()) {
+            return;
+        }
+
+        let inner_ty = generics.type_at(0);
+        let Ok(TyAndLayout { layout, .. }) = cx.tcx.layout_of(cx.param_env.and(inner_ty)) else {
+            return;
+        };
+
+        if layout.size().bytes() >= self.maximum_size {
+            return;
+        }
+
+        if implements_trait(cx, inner_ty, copy_def_id, &[]) {
+            span_lint_and_help(
+                cx,
+                COPY_REFCELL,
+                span,
+                "std::cell::RefCell used with a type that implements Copy",
+                None,
+                "replace the usage of RefCell with Cell, which is simpler and zero-cost",
```

Comment:
Bit against saying "zero-cost" here, it is correct but some interpret this as free which could lead to misinterpretation. Copying could be slower too (but not when linted). Beyond that I think we should mention bookkeeping being unnecessary + using backticks

Reply:
Done.